### ISSUE TITLE
Help test `dprint`

### ISSUE THREAD

**Issue by luisherranz** [WordPress/block-hydration-experiments#16]
@gziolo proposed fully adopting prettier (and dropping the WordPress spacing standard) in January: https://make.wordpress.org/core/2022/01/05/proposal-changes-to-javascript-coding-standards-for-full-prettier-compatibility/

Even though the proposal received a lot of support, @ntwb [said in Slack](https://wordpress.slack.com/archives/C5UNMSU4R/p1648555187344279?thread_ts=1647357562.940759&cid=C5UNMSU4R) that the consensus at the moment is a "No" from the core team. _We've been unable to locate where that core team conversation took place, and I'm not sure why Gutenberg contributors are not considered Core contributors, but the proposal is currently stuck because of that._

I'll keep trying to contact @ntwb and see if [he can finally share more information](https://wordpress.slack.com/archives/C5UNMSU4R/p1648555570050649?thread_ts=1647357562.940759&cid=C5UNMSU4R) about how, where and why that decision was made, but until then, I was testing `dprint`, as an alternative, just in case `prettier` doesn't finally work out.

As I'm not the only one using this repository anymore, it'd be great to know your impressions with `dprint`. My main issue right now is that [you cannot use a local node_modules installation](https://github.com/dprint/dprint-vscode/issues/33), and therefore people need to figure out how to install it globally. Also, I was unable to make the VS Code extension work by default with the global installation of `dprint` and I had to [manually add a `dprint.path`](https://github.com/dprint/dprint-vscode/issues/28) in the VS Code settings.

If you've also tested it out while working on this repository, please share your impressions 🙂

**Comment by luisherranz** [comment]
> My main issue right now is that you cannot use a local node_modules installation, and therefore people need to figure out how to install it globally. Also, I was unable to make the VS Code extension work by default with the global installation of dprint and I had to manually add a dprint.path in the VS Code settings.

I was able to fix those issues, which were happening because my local folder contains spaces: https://github.com/dprint/dprint-vscode/issues/33

Actually, [support for local `node_modules` was added three days](https://github.com/dprint/dprint-vscode/commit/4ae6c6874c3f3040423fccb02f7f49971f62dd01) ago 😄 🎉

**Comment by luisherranz** [comment]
I've been carefully reading the [WP JavaScript coding standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/) and trying to match them with [dprint options](https://dprint.dev/plugins/typescript/config/).

You can see the result [in the playground](https://dprint.dev/playground/#code/PTAEEkDsBMFNIC4EMEEsD2lQHdUIBajIBGAzgHQBQIoAtPdWAEQCupsAKkmUwFxEAnFrEqUAxplIJQAD1ABeUAG9KoNaCT8AjDt179BvQBpV64vxXqroMfwBMDx0+cvXTk9fXR+AZj-+AwKDggNM1AF9KSMZQADl0HHw8WFIAByQxWA1pAiz4aFB0ADNQABtUSCz0AUKsYlKkSABrMoqUqnFJaTlFLVAAbjC1TsgpWQVQO09rURoAGTbSUFJ8dBZSgrYWJFLSgE9QYizIBNLMAHNYGoJG0AAOAAYbfCQBDIQr0iMNGGXV9YKJ2ksBkmVg0BiWgeTwAFBIWIgKuciNwlkglgAWZbpTKkACU5FAHCSS1QaNAgBwCUjFBCAXAJQEJSrBvsQWNIzpBkeVKqQYpdKm9dgcKtBUGIUFkEQJYEhoNwmYUaqLqQJzo1UAAvcE2dBwKg0ei0GJMbmwADqqGgBD49weogko26EyQLtdbvdHs9XqQoAA1IdiIGg8GQ6Gw+HiIMYqgisBYKV2MAitVgNgkkzgAgBAd6ugxE0lis1hsNKVsEg9ks2FliG9cd9GgUduXK6BzglMDEALbrNCpBWmiigGKGmKgVjsABCdZSNqYzYrpCYfvHpCRTMnur2AAV0Gu0Jg55UZAgFpUmKIY6AYeN5HfQH08RoJlpBvH2M-FHYo8mBDemU6igPP04wADwPsBsi+v6T5KBoN5yE+0RpqgTIwnI4FQrB8HjEhohZgccEyDePpITYKBiIQMJXAI1TYXsJGgHhMQAKqQK8BxpLAYioDstCUa87xXIUqRXCg1RLNR5DnOQ3zQd89BPj2YxAqALwAG5ZGkGTHCCOToDEuSoDU6CiW8MAdA6Yw9M+0GDFZgF0LQGhRjQACCkAHA2vwgcp0iqRpWSpNKmSipy2I6R0xHPiyoDfGITEDLMYAeQcIHVk25JSBKXbwDkVxdhU4k1H5hxZCgRD4HkvzFJVWSmpZXSfg+IEOYcEzfiM1kTDCT7yAAfMozWvu1X4gZERQImIB5YMR2E+r0IHEB143JaAqWgPwSBFB8NQ+sFplXAgBzsbloAVD8hTEAAVtx0hwEUFR4BgWClQFSCaaAwXceCSIxNpmSNY6t7KEMPraB4ZhWPw6h2CY0Q0MSWQAPw-AU-AXT6u3sdmOowM9mA7KApWBRFmS1Ic6AEMslrtPaTU2T6g0PMjWi8J1MTxKAj27MJAMpOdWCwF2qTHXjUhCNNkmwNJsnKOE3wANoALrfEUkC9QSojRXigyTZA00vbIN6wZEbU2YroDK-ZDMTHB4SuSlnmgAAhKAlRqjNIliQg1R-MWBSkz6ya7OguDhfz+pgKONBc6ZXt7LA+mgNACQEGSHRXjeLtyMziWWOoC0tVEa1ubs3NTV7xC6qgAuvFkIp5dqxAHJgWQkN8sCaVgV65KAeXGVkPMN7yNDYG8qSiQUmM2GcpAsNKHT64bmDG7BlBF6+JeUDCy9e3NoOb4M4R4r1a0API3XdaK-K8bytuKdRZHAYgNNKBSrz6a6cgObTnSUuQDj139gIBAUcnJGhoEwb6RQrgAGV1ywDPLAG0WZhD02BjZOCYMIZDHMJMeGNtMETEVmDSGahiCUGtmtM0VUsC3HQFfaaioNACHvudJYvtTgXCIAkR60hV5t1aJUTuGRCC5S7EcEqbBpBHBiP2HSH8e4ICWGHLApo0b9zEcI6qcACgt2fBILsXYkDgJjswLMSBUJIgAMLoGMeiOcmB9gAFleyoGQReLqjkC5qE0A+QwgSgk6BZPYNw4TwnkJsL4EIsTYlRCId1RQpCrDejSekpALIIzZJyVEsQ+SCmFKKcUkpBSqGO3WmWRcgtX4sDgP3E8bwyZ11ogiAo8ZhZ5Rvk2VULBcqIAoCOBgUD+ZuVaTAVBQgRDeJBorTJhxqHDmIvM4gutVAzMZorAA5MQLZ1tQCxypssFgk9qgfAKInMB6zzbOkVg8fZhzpDz1OaA7UlzM4lHQqAAAZN868BxfmgA1HiJ8R8DlgDgbAYWSxyhNHKocFgyJV4AANRnjOgMikSM0Og0BYh+dghUJAciHLELuwlpT7Apm5Nk9iUBilAJCwl6AOQQFGEdI2MI3JwPAFrDZzo+WKF6goQacFN6gARmALgZB-YAjKqADK3M-aN0QHSzAhIACi3d-51X7oiaU3NUIjzxsgNo09IAxB9K-PcC8sgwlQOQGW3lzrGN+hKSlFR1LoDhQUPeL08TfD7g6D4AzCgAJeAgGIvrV5FllUcQWcBEDN1bpUFExAOi70rhy5868kC9T1pm1exAhUqASotEuJ8EJMQqTGTuCZmSKoEN8FCTJvIFAIocM4+ZCz-BLAuVs1ZDgzi+JovtSx2wxFXj2UofZf48kJEjb2kAkSDp0jKkscbV591IEgM6Gj0R1UjQWrAD0nozQDVVPGooZo7G+H7PuZxTLzovVa7+yJayrpjRsGIG6sD3r-qKEKCBKUhzOOHZE970RPOQB8fpAiw01k7U0KgWdcLDTfHW-+O9UNF06u+eFigfBRhQ4hSgvi0PRF-NeGyI1c4QRAjIaCeJSNgwmDIfoyF0y2to1hYcZGi5sYo0etezHUkvnYzQuhz5sowbyhw3h3Dwq+wNQIrAQjTTfA0bWGUBZiYyMKPkheGgdpXAtapsyxUzHDOYIdN4vsBC7n3C9Oc27cqeIwUkjQnoVwRhXKUsQQx-TQCC8F4LK4oXhYiyuIo0WYsxYC22c4iWkuJYqcg7tAc5W1i9fAQWSmzjnDFETc4rTUikhKHgZ1B1NJLGlLKbg1jjrfHnpRC1hZ+x4DQOFGUlEGnfVIGuQRJQsZXBxq3czdnahKbwKo7A6i2id27tGEo1dqa4HLvwimX9EE6KBh5jJLofl-JySGUAqM-N+c2inEL12btBcSY5fbPpAXHeDKdmw52SmbSGLdn7wWKm0Jy5azAV6XpEzJPJsoPClMbcESm9TWies2cbKG58+XCulAneNpVv6L3V2ZTKIWMg+sDdenppAk8ZQmWUTNubIj41NyUSIGgekmlMk0qUbmtEuw6sOku8KyOX3LvSNKRAVV2CDINFZ8cxAirZnVUT6U-WXrkFNNuK48vidOf4Gg6ZKHHvekOwGF74ZDcfbN-nIYc1j4ScB88KxqmSi5QILqciuxSScPQAp8DfCKuw-qvNuq+rSpyJoEI8U5czI6KbXgQgfdHoCDGOHjnn9liwGFxKHRnP7E86YTkKq3Yk6rGgGifVSftRqogPB4mheXcCU5ALQNmAPgnidSzn0bP4yhujDAPKqrSdjDjRlSzkDmASKkRrxXJOVdtDVwICfKQSeTOEH6Y0Y-1cK4X8rmB8DEGeP4EUHY7BtbkH156G8j51DkGN9km8UwmPkDN35m8Pgq3ayGCf0-7pz9MSGFYK-1-wxb9EoaA7FE0Tx6cVUvZwdVJnlUgzlwQqA-9H9Sln9X8Yg3IldzhIBYMlhcB04ccsh1JXhlhoNOlEA119F-cpBy8e58Z1JLRtgOdBwTAx5OMU5uI34+8lhH45Vis1gp5NsaYf5-dKhdtpAVkTAxBvgChehO4QJHljkXlzlQB3k1okYg89M40tthDmkyoEBsAoUCCdQXVyDfY4UsAYQQBQVfhG8TCI0aBm8rlJcR9xwjFYNkFyBfxMg4EcRYA3JjMBA4EGgVhZxtcplko7E7DiYyRX1mkcUwBIicDjCaxYBfx25n1dR24EgUIetAEGRUirhW0KDSwWwlgQ8wBvpQok1nx6hGgWgGp3NxCxMYhEjZNuA1hZEnI4gjkkAaUTE0Ak8gEgttQ28dEtMkAWhN1TIVCk4Zlloxo1o8UtJEFaAABHFgKmAWSjCWZdcoXaQ-YfY0dYzYuBY6JkOcUdBBYQrxNqIuJgHYzkJgf7STL+LMZdINe3QsbbY4j4YdPIyobUJTONFIcUAQvAwgJAb9DIAsYIqiAAHTxF4CGRcKYB+NgFOL2HOP4HnCqUrCuPOHOMaOanuLePCnBNAC2VfQVDRNIC2SeJtywDYGXVIDwJ62kzIJUQgOkHFHxVIKSKERIBwBjwulyH+lZPwA6BZLwEokrXXh5KyDZg3jzUoHGKaHsnRCyDsCRKLTWVVMGAel6OnSRLEGVIlVAEvlumYQOjMjQAFk-UoI0HyQX21HoJ9FTn8ipj72+ARCZH6x1ThWFDRBiCIIEF4nqCqD2hINDMU3DRwAy1Ui0NACIPKGngTTQEeiuCihP0GBkDmWoVzK2SQD2XkLAHiHEL6LpUGMEEaFIF-FyjbQSGRRkBP0xVUILKQFoCLOoRiHUK0h7QdJ9EqGwEzwqj7nyBR260IGHnICAA/config/N4KABGBEA2CWB2BTA6rAJgFwBaQFxgA4AGAGnCgTUXg1Ux3wBYyJIBXAZ0QBUBDAIw54wGAE5tELKFwC2sAMIB7aIvhD8kAA6jEAM0SjIUyAEc2ijIgDKGAJ7REwyL2gB3XrY5WEAcwdHyU3NLAAVRRU11KG1ELlEAN0djJFcAGQREAGlKJ2hdANZORAAhUV4AY1inF3dPAqh+MsqQxQ5YDFhVJw5eGUR0pABVeAcODgAJXngfX3rINumHYsU0Wxa2jq6NJAAPDAGkwN2MJRpw6AAxFVd19s74J2ODubFeWDhppRkZXijIVWgtgAsmxoB1nsYIgZeBhFKJbpsHhoen0IYEYvpRJNprN8LoXFxjBiDN5Fv0MsIxBJjH0ZPwDABRHYxMb3AB0xNEpL85KQwnx0EJgV4onCrgubHg5URbKKIRF1GwsSqGh+CAwbwexn4CBFtiZLLaqjZH0QIUZzJ0rK2InEkkCACsODs2WYLNY7P4NJyACKKNj8fzGJ0umSg8EZeU6NROTlzWn00QGq1G+AmyMWw33Sl24x2TTkyxlaBsriaEUwuHZeBobqIORKFRawLUNgyH2IcrQCsyhMk8vlXFQNU0TVzDgDxAAQXCktr+Cp9tYE4q1jYov9Nd8YShog6Ktt1MCin4Ds7GGTsVTpcnVnXs630x3Bb3sAPi8hp-P8owRbTK8qO8Nznbdwhffcog-QJ80LaESwAtdgMfHxnwMCCcyPVgdXgPVL2tf9b3vTc0F8Yp2lcWAuCnGsZ3aLA+g6coAHkX0rQwF1zQJykUb5FQONldDhQDJynXQiysbsOCwd9ONYbi1CpaU4RvVdij0OEzQVGhpK4KIBSFOTVA4RTYVEbhbALFTKlEosADlEFcTJEFsVw4XnMB9KXKB5OM8RpW8HwcIwddECs6cxIMezHOc1zRHczzjDQRRkCwd4PRhetFTCmyDBStKnJctyMK8yBEGZOEMA7Lse3ZBCgIfEjpls3pEDQA0Ksg2SoEE0QAEl4BsDK+hobKItEC4qxioqOMw7q4SY3RBssYaMFGosJtEArYvcqDWB6pbMpGhCcvGybCri4rjH2jVlqy46xqsesFGUIzLsCXRJWle4qu7MoZQQtSes06MlV0-kCRKj6pURPDrwB9SdCjRUdIPBL3s+mHLSvWqRLGiVofuLbpo8iHjB8RAMCncpKjGZT4aBpHtOVPTScCWBFpuw7Vvuotet0ImLpmkrYBkTQKp+mrjTqoiQKalq0F60WOre1gQwZBwVrCwGNMevJ5BUBYfD4HwJN+BhDxKtWsfw04NQyUQwvq4jfFh7M8VZ1gGKwFYtYR4HkeZ8HBRKrhKepq86cnbXEa00HUY9qANR8cm0G4etNG7SxfaB9IixcFXE4s6d4HgCwYRx1S-cbG00dYGCpzGNCK8qaPEFdm1dqgVxUocA7NZ53Ke6yKbBYt4xYECjTbJWRAvhWtO9icNBtHVABaCfS50OYN40i40rnxUF4wJeV5odfJ50VfdDSyAQAAXxAIA/language/typescript).

**TLDR;**\
Most things work with the current options.

- I've added same the options to this repo and merged this PR: https://github.com/luisherranz/block-hydration-experiments/pull/17
- I've merged/rebased the rest of the PRs.
- I've made a PR to fix a bug in `dprint`: https://github.com/dprint/dprint-plugin-typescript/pull/368

**Things that don't work**

- Spaces when accessing properties/items of objects/arrays

  ```js
  const x = a["b"];
  const x = a[0];

  // It should be:
  const x = a[ "b" ];
  const x = a[ 0 ];
  ```

  _It should be easy to add a new option for this if they agree._

- Context indentation in chains

  ```js
  x.
    a( 1 )
    .aa( 2 )
    b( 3 )

  // It should be:
  x.
    a( 1 )
      .aa( 2 )
    b( 3 )
  ```

  _I don't think this will ever work because it can be inferred from the code._

- Using individual lines in variable declarations.

  ```js
  const a, b, c = 1, d;

  // It should be:
  const a, b,
    c = 1,
    d;
  ```

  _I'm not sure how easy it would be to add this functionality. It probably can be done, but not sure if it'll be worth the effort._

- Automatically adding a new line above comments

  ```js
  const a = 1;
  // Comment.
  const b = 2;

  // Instead it should be:
  const a = 1;

  // Comment.
  const b = 2;
  ```

  _This is not enforced or automatically transformed, but the developer can add the new line manually, so it's not a deal-breaker._

  _I guess this can be added, although I'm not sure if it will conflict with other rules, like no new line at the start of a block._

- Automatically transform object props to dot notation whenever possible

  ```js
  x['a'];

  // Instead t should be:
  x.a;
  ```

  _This is not enforced or automatically transformed, but the developer can add the new line manually, so it's not a deal-breaker._

  _I guess this can be added to `dprint`. Prettier already does it._

**Bugs that affect WP coding standards**

- Spaces on functions without arguments

  ```js
  x(  ) === true

  // It should be:
  x() === true
  ```

  _I've already created a [PR](https://github.com/dprint/dprint-plugin-typescript/pull/368) to fix it._

- No spaces in the inner parenthesis

  ```js
  if ( x && (y && z) ) a = 1;

  // It should be:
  if ( x && ( y && z ) ) a = 1;
  ```

  _It should be easy to fix._

---

You can see the configuration in the [`dprint.json` file](https://github.com/luisherranz/block-hydration-experiments/blob/main/dprint.json).

```json
{
	"useTabs": true,
	"lineWidth": 80,
	"typescript": {
		"quoteStyle": "alwaysSingle",
		"useBraces": "always",
		"singleBodyPosition": "nextLine",
		"operatorPosition": "sameLine",
		"preferSingleLine": true,
		"memberExpression.preferSingleLine": false,
		"binaryExpression.linePerExpression": true,
		"memberExpression.linePerExpression": true,
		"spaceAround": true
	}
}
```

**Comment by gziolo** [comment]
@luisherranz, I think you should be comparing with what `wp-prettier` does with the current config from `@wordpress/prettier-config`:

https://github.com/WordPress/gutenberg/blob/trunk/packages/prettier-config/lib/index.js

I'm not 100% sure that all WP JS coding standards can be achieved with automated formatting so we tried to bring it as close as possible. We might miss some changes as well in the process.

It probably would be best to run `wp-prettier` with the config I shared on the files formatted with dprint to learn about differences.

**Comment by luisherranz** [comment]
> I think you should be comparing with what wp-prettier does with the current config from @wordpress/prettier-config

Oh, ok. I'll try to match the options you are currently using with `prettier` that are not present in the coding standards, like `jsxBracketSameLine: false` and `arrowParens: 'always'`. Thanks 🙂

> It probably would be best to run wp-prettier with the config I shared on the files formatted with dprint to learn about differences.

I'll do that as well, good idea. 

I think there are aesthetic differences between both in the formatting, but nothing that affects the coding standards. 

If we ever move Gutenberg to `dprint`, we'd need to do one of those "git blame ignore" commits, though.

**Comment by luisherranz** [comment]
For the record, I've been trying to set up `@prettier/plugin-php` with VS Code and hasn't been straightforward either. I still haven't managed to get it working with markdown `php` snippets 😕

**Comment by ockham** [comment]
I'm probably Doing It Wrong, but FWIW, the VS Code extension doesn't seem to work for me:

![dprint](https://user-images.githubusercontent.com/96308/174099314-a9fffc52-853c-4188-836b-a2f9a716e87b.gif)

Maybe we can add auto-formatting to the `git` commit hook? I think GB does something like that (or maybe just linting).

**Comment by luisherranz** [comment]
Oh, that's a shame 😩

Does the VS Code prettier extension work for you in that project?

**Comment by luisherranz** [comment]
I'm going to stop this experiment and move back to Prettier to start testing the [`prettier-php` plugin](https://github.com/prettier/plugin-php) as well.

**Comment by luisherranz** [comment]
Done in https://github.com/WordPress/block-hydration-experiments/pull/43. 

I'll add `prettier-php` in a future PR.
